cluster/kubernetes/sync.go

```diff
@@ -2,31 +2,308 @@ package kubernetes
+		} else {
+			errs = append(errs, cluster.ResourceError{ResourceID: res.ResourceID(), Source: res.Source(), Error: err})
+			break
+		}
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.muSyncErrors.RLock()
+	if applyErrs := c.applier.apply(logger, cs, c.syncErrors); len(applyErrs) > 0 {
+		errs = append(errs, applyErrs...)
+	}
+	c.muSyncErrors.RUnlock()
+
+	if c.GC {
+		orphanedResources := makeChangeSet()
+
+		clusterResources, err := c.getResourcesInSyncSet(spec.Name)
+		if err != nil {
+			return errors.Wrap(err, "collating resources in cluster for calculating garbage collection")
+		}
+
+		for resourceID, res := range clusterResources {
+			actual := res.GetChecksum()
+			expected, ok := checksums[resourceID]
+
+			switch {
+			case !ok: // was not recorded as having been staged for application
+				c.logger.Log("info", "cluster resource not in resources to be synced; deleting", "resource", resourceID)
+				orphanedResources.stage("delete", res.ResourceID(), "<cluster>", res.Bytes())
+			case actual != expected:
+				c.logger.Log("warning", "resource to be synced has not been updated; skipping", "resource", resourceID)
+				continue
+			default:
+				// The checksum is the same, indicating that it was
+				// applied earlier. Leave it alone.
+			}
+		}
+
+		if deleteErrs := c.applier.apply(logger, orphanedResources, nil); len(deleteErrs) > 0 {
+			errs = append(errs, deleteErrs...)
+		}
+	}
+
+	// If `nil`, errs is a cluster.SyncError(nil) rather than error(nil), so it cannot be returned directly.
+	if errs == nil {
+		return nil
+	}
+
+	// It is expected that Cluster.Sync is invoked with *all* resources.
+	// Otherwise it will override previously recorded sync errors.
+	c.setSyncErrors(errs)
+	return errs
+}
+
+// --- internals in support of Sync
+
+type kuberesource struct {
+	obj        *unstructured.Unstructured
+	namespaced bool
+}
+
+// ResourceID returns the ResourceID for this resource loaded from the
+// cluster.
+func (r *kuberesource) ResourceID() flux.ResourceID {
+	ns, kind, name := r.obj.GetNamespace(), r.obj.GetKind(), r.obj.GetName()
+	if !r.namespaced {
+		ns = kresource.ClusterScope
+	}
+	return flux.MakeResourceID(ns, kind, name)
+}
+
+// Bytes returns a byte slice description
+func (r *kuberesource) Bytes() []byte {
+	return []byte(fmt.Sprintf(`
+apiVersion: %s
+kind: %s
+metadata:
+  namespace: %q
+  name: %q
+`, r.obj.GetAPIVersion(), r.obj.GetKind(), r.obj.GetNamespace(), r.obj.GetName()))
+}
+
+func (r *kuberesource) Policy() policy.Set {
+	return kresource.PolicyFromAnnotations(r.obj.GetAnnotations())
+}
+
+// GetChecksum returns the checksum recorded on the resource from
+// Kubernetes, or an empty string if it's not present.
+func (r *kuberesource) GetChecksum() string {
+	return r.obj.GetAnnotations()[checksumAnnotation]
+}
+
+// GetSyncSet returns the sync set name recorded on the the resource
+// from Kubernetes, or an empty string if it's not present.
+func (r *kuberesource) GetSyncSet() string {
```

Comment:
This is dead code.

Reply:
True -- because we just select it directly from the API now. Well spotted!